File: assets/setup-ui/js/stores/scopeForm.js
document.addEventListener('alpine:init', () => {
  Alpine.store('scopeForm', {
    loading: false,
    saving: false,
    error: null,
    saveError: null,
    providers: [],        // [{ id, display_name, form_spec, current_values, question_extras }]
    answers: {},          // { provider_id: { field_key: value } }
    fieldErrors: {},      // { provider_id: { field_key: error_key } }
    questionExtras: {},   // { provider_id: { question_id: { placeholder, docs_url, group } } }
    currentStep: 0,       // 0-indexed; maps to providers[currentStep], or review if === providers.length
    manualSteps: [],      // [{ provider_name, steps }] — populated after save when manual portal steps are needed

    get totalSteps() {
      return this.providers.length + 1; // +1 for review step
    },

    isReviewStep() {
      return this.currentStep === this.providers.length;
    },

    currentProvider() {
      return this.providers[this.currentStep] || null;
    },

    reset() {
      this.currentStep = 0;
      this.saveError = null;
      this.fieldErrors = {};
      this.questionExtras = {};
      this.manualSteps = [];
    },

    async refresh() {
      this.loading = true;
      this.error = null;
      const scope = Alpine.store('scope');
      try {
        const data = await window.api.get(
          '/api/scope/form?tenant=' + encodeURIComponent(scope.tenant) +
          '&env=' + encodeURIComponent(scope.env) +
          '&team=' + encodeURIComponent(scope.team)
        );
        this.providers = data.providers || [];
        // Initialize answers from current_values for each provider.
        const init = {};
        const extras = {};
        for (const p of this.providers) {
          init[p.id] = { ...(p.current_values || {}) };
          extras[p.id] = p.question_extras || {};
        }
        this.answers = init;
        this.questionExtras = extras;
        this.fieldErrors = {};
        this.currentStep = 0;
      } catch (err) {
        this.error = err;
      } finally {
        this.loading = false;
      }
    },

    /// Get a single extra metadata field for a question.
    /// Returns null when extras are absent for that provider or question.
    getExtra(providerId, questionId, field) {
      const pe = this.questionExtras[providerId];
      if (!pe) return null;
      const qe = pe[questionId];
      if (!qe) return null;
      return qe[field] || null;
    },

    fieldValue(providerId, fieldKey) {
      return (this.answers[providerId] || {})[fieldKey] || '';
    },

    setFieldValue(providerId, fieldKey, value) {
      if (!this.answers[providerId]) this.answers[providerId] = {};
      this.answers[providerId][fieldKey] = value;
      // Clear field-level error on edit.
      if (this.fieldErrors[providerId] && this.fieldErrors[providerId][fieldKey]) {
        delete this.fieldErrors[providerId][fieldKey];
      }
    },

    /// Evaluate whether a question should be visible based on its visible_if expression.
    /// Returns true if the field should be shown.
    ///
    /// Supports two formats:
    ///   "field_name==value"  → show if answers[field_name] === value
    ///   "field_name"         → show if answers[field_name] is truthy (non-empty, non-null, non-"false")
    isVisible(providerId, questionId) {
      const vis = this.getExtra(providerId, questionId, 'visible_if');
      if (!vis) return true; // No condition → always visible.

      const pa = this.answers[providerId] || {};

      if (vis.includes('==')) {
        const [field, expected] = vis.split('==', 2);
        const actual = String(pa[field] || '');
        return actual === expected;
      }

      // Truthy check: non-empty, non-null, not literally "false".
      const val = pa[vis];
      if (val === undefined || val === null || val === '' || val === 'false' || val === false) {
        return false;
      }
      return true;
    },

    /// Display a value for the review step: masked if secret, raw if not.
    displayValue(providerId, fieldKey, isSecret) {
      const v = this.fieldValue(providerId, fieldKey);
      if (v === undefined || v === null || v === '') return '(empty)';
      const s = String(v);
      if (isSecret) {
        if (s.length <= 4) return '••••';
        return '••••' + s.slice(-4);
      }
      return s;
    },

    /// Whether the user can advance to the next step (required fields filled).
    canAdvance() {
      if (this.isReviewStep()) return this.canSave();
      const provider = this.currentProvider();
      if (!provider) return false;
      const pa = this.answers[provider.id] || {};
      for (const q of (provider.form_spec && provider.form_spec.questions) || []) {
        if (!q.required) continue;
        if (!this.isVisible(provider.id, q.id)) continue; // Skip hidden fields.
        const v = pa[q.id];
        if (v === undefined || v === null) return false;
        if (typeof v === 'string' && v.trim() === '') return false;
      }
      return true;
    },

    nextStep() {
      if (this.canAdvance() && this.currentStep < this.providers.length) {
        this.currentStep++;
      }
    },

    prevStep() {
      if (this.currentStep > 0) {
        this.currentStep--;
      }
    },

    async save() {
      this.saving = true;
      this.saveError = null;
      this.fieldErrors = {};
      const scope = Alpine.store('scope');
      try {
        const data = await window.api.post('/api/scope/form', {
          scope: { tenant: scope.tenant, env: scope.env, team: scope.team },
          by_provider: this.answers,
        });
        // Store any manual post-setup instructions returned by the backend.
        this.manualSteps = data.manual_steps || [];
        // Refresh overview to reflect new state.
        if (Alpine.store('overview')) {
          await Alpine.store('overview').refresh();
        }
        console.info('[scopeForm] saved', data);
        return data;
      } catch (err) {
        this.saveError = err;
        // If err.fields is populated, distribute to fieldErrors by provider.
        if (err && err.fields && typeof err.fields === 'object') {
          for (const k of Object.keys(err.fields)) {
            // Field keys are formatted as "provider_id.field_key".
            const dotIdx = k.indexOf('.');
            if (dotIdx > -1) {
              const pid = k.substring(0, dotIdx);
              const fk = k.substring(dotIdx + 1);
              if (!this.fieldErrors[pid]) this.fieldErrors[pid] = {};
              this.fieldErrors[pid][fk] = err.fields[k];
            }
          }
        }
      } finally {
        this.saving = false;
      }
    },

    /// Whether all required fields across all providers have non-empty values.
    canSave() {
      for (const provider of this.providers) {
        const pa = this.answers[provider.id] || {};
        const questions = (provider.form_spec && provider.form_spec.questions) || [];
        for (const q of questions) {
          if (!q.required) continue;
          if (!this.isVisible(provider.id, q.id)) continue; // Skip hidden fields.
          const v = pa[q.id];
          if (v === undefined || v === null) return false;
          if (typeof v === 'string' && v.trim() === '') return false;
        }
      }
      return true;
    },
  });
});

File: src/ui/api/scope_form.rs
//! Unified scope form endpoints: GET/POST `/api/scope/form`.
//!
//! These endpoints replace the separate wizard and secrets views in the SPA.
//! A single GET returns every provider's FormSpec + current values pre-filled
//! from the secrets store. A single POST validates all answers and persists
//! them directly via `persist_qa_secrets`.
//!
//! Security invariants:
//! - GET returns raw current secret values because the user is editing them.
//!   This is acceptable because the endpoint is bearer-auth + origin-checked.
//!   Values are wrapped in `Zeroizing<String>` server-side and never logged.
//! - POST accepts plain values, validates per FormSpec, persists via
//!   `persist_qa_secrets`. `pending_mutations` is NOT set because this is a
//!   direct persist — no rebuild is needed.
//! - No raw values appear in any log entry. Audit entries record only
//!   identity fields (provider_id, key, scope).

use std::collections::HashMap;
use std::sync::Arc;

use axum::Json;
use axum::extract::{Query, State};
use axum::response::IntoResponse;
use greentic_secrets_lib::SecretsStore;
use serde::{Deserialize, Serialize};
use serde_json::{Value, json};
use tracing::info;
use zeroize::Zeroizing;

use crate::ui::api::error::{ApiError, FieldError};
use crate::ui::state::{AppState, ScopeKey, validate_scope};

// ── Request / response types ──────────────────────────────────────────────────

#[derive(Debug, Deserialize)]
pub struct ScopeQuery {
    pub tenant: String,
    pub env: String,
    pub team: String,
}

/// Response payload for `GET /api/scope/form`.
#[derive(Debug, Serialize)]
pub struct ScopeFormResponse {
    pub scope: ScopeKey,
    pub providers: Vec<ProviderFormEntry>,
}

/// Extended metadata for a single question, sourced from `assets/setup.yaml`.
///
/// These fields are not part of the `qa_spec::FormSpec` schema; they come from
/// the provider's legacy setup spec. Missing fields are omitted from JSON.
#[derive(Debug, Clone, Serialize, Default)]
pub struct QuestionExtras {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub placeholder: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub docs_url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub group: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub visible_if: Option<String>,
}

/// One provider's FormSpec + current values (raw, for editing).
#[derive(Debug, Serialize)]
pub struct ProviderFormEntry {
    pub id: String,
    pub display_name: String,
    pub form_spec: qa_spec::FormSpec,
    /// Current values from the secrets store for each question id.
    /// Only keys with an existing value are included (no empty placeholders).
    /// Values are raw strings — this is intentional for the edit form.
    pub current_values: HashMap<String, String>,
    /// Extra per-question metadata from `assets/setup.yaml` (placeholder,
    /// docs_url, group). Keyed by question id. Empty when setup.yaml is absent.
    pub question_extras: HashMap<String, QuestionExtras>,
}

/// Request body for `POST /api/scope/form`.
#[derive(Debug, Deserialize)]
pub struct PostScopeFormBody {
    pub scope: ScopeKey,
    /// Answers keyed by provider_id → { field_key: value }.
    pub by_provider: HashMap<String, Value>,
}

/// Success response for `POST /api/scope/form`.
#[derive(Debug, Serialize)]
pub struct PostScopeFormResponse {
    pub success: bool,
    pub providers_saved: usize,
}

// ── GET /api/scope/form ───────────────────────────────────────────────────────

/// Return the unified scope-form payload: all providers with their FormSpec
/// and current values pre-filled from the secrets store.
pub async fn get_scope_form(
    State(state): State<Arc<AppState>>,
    Query(q): Query<ScopeQuery>,
) -> Result<impl IntoResponse, ApiError> {
    let scope = ScopeKey {
        tenant: q.tenant.clone(),
        env: q.env.clone(),
        team: q.team.clone(),
    };
    validate_scope(&scope, &state.bundle).map_err(|e| ApiError::validation(&e.code, &e.key))?;

    info!(
        action = "get_scope_form",
        tenant = %scope.tenant,
        env = %scope.env,
        team = %scope.team,
        providers = state.provider_forms.len(),
        "scope form requested"
    );

    // Open the secrets store once and read all relevant values.
    // We do this in a blocking task because the dev store does sync I/O under
    // the hood (even though the API is async).
    let bundle_path = state.bundle.path.clone();
    let provider_forms = state.provider_forms.clone();
    let scope_clone = scope.clone();

    let providers: Vec<ProviderFormEntry> =
        tokio::task::spawn_blocking(move || read_provider_form_entries(&bundle_path, &scope_clone, &provider_forms))
            .await
            .map_err(|e| {
                ApiError::internal("scope_form.task_panic", "ui.error.internal")
                    .with_params(json!({ "message": e.to_string() }))
            })?
            .map_err(|_| ApiError::internal("scope_form.read_failed", "ui.error.secrets_list_failed"))?;

    Ok(Json(ScopeFormResponse { scope, providers }))
}

/// Read FormSpec + current values for every provider (runs in blocking context).
fn read_provider_form_entries(
    bundle_path: &std::path::Path,
    scope: &ScopeKey,
    provider_forms: &[crate::ui::state::ProviderFormData],
) -> anyhow::Result<Vec<ProviderFormEntry>> {
    // Build a synchronous runtime for the async store reads.
    let rt = tokio::runtime::Handle::try_current()
        .ok()
        .map(|_| None::<tokio::runtime::Runtime>);

    // We're already inside spawn_blocking so we can create a mini rt to drive
    // the async store operations.
    let mini_rt = tokio::runtime::Builder::new_current_thread()
        .enable_all()
        .build()?;

    let _ = rt; // suppress warning

    let store = crate::secrets::open_dev_store(bundle_path)?;

    let mut entries = Vec::new();

    for pf in provider_forms {
        let mut current_values: HashMap<String, Zeroizing<String>> = HashMap::new();

        for q in &pf.form_spec.questions {
            let uri = crate::canonical_secret_uri(
                &scope.env,
                &scope.tenant,
                Some(scope.team.as_str()),
                &pf.provider_id,
                &q.id,
            );

            // Drive the async get() on the mini runtime.
            let bytes = mini_rt.block_on(store.get(&uri));
            if let Ok(b) = bytes
                && !b.is_empty()
                && let Ok(text) = String::from_utf8(b)
            {
                current_values.insert(q.id.clone(), Zeroizing::new(text));
            }
        }

        // Convert to plain HashMap<String, String> for JSON serialization.
        // The Zeroizing wrapper drops at end of this scope.
        let plain_values: HashMap<String, String> = current_values
            .iter()
            .map(|(k, v)| (k.clone(), v.as_str().to_owned()))
            .collect();

        // Load extended question metadata from setup.yaml inside the pack.
        // Failures are non-fatal — missing setup.yaml produces an empty map.
        let mut question_extras: HashMap<String, QuestionExtras> =
            match crate::setup_input::load_setup_spec(&pf.pack_path) {
                Ok(Some(spec)) => {
                    spec.questions
                        .into_iter()
                        .filter(|q| !q.name.is_empty())
                        .map(|q| {
                            let extras = QuestionExtras {
                                placeholder: q.placeholder,
                                docs_url: q.docs_url,
                                group: q.group,
                                visible_if: None,
                            };
                            (q.name, extras)
                        })
                        .collect()
                }
                _ => HashMap::new(),
            };

        // Merge visible_if expressions from FormSpec questions into question_extras.
        for q in &pf.form_spec.questions {
            let vis = q
                .visible_if
                .as_ref()
                .and_then(crate::ui::api::wizard_engine::expr_to_string);
            if vis.is_some() {
                question_extras
                    .entry(q.id.clone())
                    .or_default()
                    .visible_if = vis;
            }
        }

        entries.push(ProviderFormEntry {
            id: pf.provider_id.clone(),
            display_name: pf.display_name.clone(),
            form_spec: pf.form_spec.clone(),
            current_values: plain_values,
            question_extras,
        });

        // Explicitly drop to zeroize secrets from this provider before moving to next.
        drop(current_values);
    }

    Ok(entries)
}

// ── POST /api/scope/form ──────────────────────────────────────────────────────

/// Validate all provider answers against their FormSpecs and persist to the
/// secrets store via `persist_qa_secrets`.
pub async fn post_scope_form(
    State(state): State<Arc<AppState>>,
    Json(body): Json<PostScopeFormBody>,
) -> Result<impl IntoResponse, ApiError> {
    let scope = &body.scope;
    validate_scope(scope, &state.bundle).map_err(|e| ApiError::validation(&e.code, &e.key))?;

    info!(
        action = "post_scope_form",
        tenant = %scope.tenant,
        env = %scope.env,
        team = %scope.team,
        providers = body.by_provider.len(),
        "scope form save requested"
    );

    // Validate each provider's answers against FormSpec.
    let mut api_err = ApiError::validation("scope_form.validation_failed", "ui.error.validation_failed");
    let mut has_field_errors = false;

    for (provider_id, answers) in &body.by_provider {
        let answers_obj = if answers.is_object() {
            answers.clone()
        } else {
            json!({})
        };

        // Unknown provider_ids are silently skipped (no FormSpec to validate against).
        if let Some(form_data) = state
            .provider_forms
            .iter()
            .find(|pf| &pf.provider_id == provider_id)
            && let Err(msg) = crate::qa::wizard::validate_answers_against_form_spec(
                &form_data.form_spec,
                &answers_obj,
            )
        {
            // Prefix field errors with provider_id.field_key so the frontend
            // can distribute them to the correct provider section.
            let field_key = format!("{provider_id}.{msg}");
            api_err = api_err.with_field(&field_key, FieldError::new("ui.error.validation_failed"));
            has_field_errors = true;
        }
    }

    if has_field_errors {
        return Err(api_err);
    }

    // All valid — persist each provider's answers.
    let bundle_path = state.bundle.path.clone();
    let scope_clone = scope.clone();
    let provider_forms = state.provider_forms.clone();
    let by_provider = body.by_provider.clone();

    let count = tokio::task::spawn_blocking(move || {
        persist_all_providers_blocking(bundle_path, scope_clone, provider_forms, by_provider)
    })
    .await
    .map_err(|e| {
        ApiError::internal("scope_form.task_panic", "ui.error.internal")
            .with_params(json!({ "message": e.to_string() }))
    })?
    .map_err(|e| {
        ApiError::internal("scope_form.persist_failed", "ui.error.execute_failed")
            .with_params(json!({ "message": e }))
    })?;

    // This is a direct persist — no rebuild is needed, so we clear pending.
    state.pending_mutations.store(false, std::sync::atomic::Ordering::Relaxed);

    info!(
        action = "scope_form_saved",
        tenant = %scope.tenant,
        env = %scope.env,
        team = %scope.team,
        providers_saved = count,
        "scope form saved successfully"
    );

    Ok(Json(PostScopeFormResponse {
        success: true,
        providers_saved: count,
    }))
}

/// Persist answers for all providers (runs in blocking context).
fn persist_all_providers_blocking(
    bundle_path: std::path::PathBuf,
    scope: ScopeKey,
    provider_forms: Vec<crate::ui::state::ProviderFormData>,
    by_provider: HashMap<String, Value>,
) -> Result<usize, String> {
    let store = crate::secrets::open_dev_store(&bundle_path)
        .map_err(|e| format!("failed to open secrets store: {e}"))?;

    let mini_rt = tokio::runtime::Builder::new_current_thread()
        .enable_all()
        .build()
        .map_err(|e| format!("failed to build runtime: {e}"))?;

    let mut saved_count = 0usize;

    for (provider_id, answers) in &by_provider {
        let answers_obj = if answers.is_object() {
            answers.clone()
        } else {
            json!({})
        };

        // Find the FormSpec for this provider (skip unknown providers).
        let Some(form_data) = provider_forms.iter().find(|pf| &pf.provider_id == provider_id)
        else {
            continue;
        };

        let saved = mini_rt
            .block_on(crate::qa::persist::persist_qa_secrets(
                &store,
                &scope.env,
                &scope.tenant,
                Some(scope.team.as_str()),
                provider_id,
                &answers_obj,
                &form_data.form_spec,
            ))
            .map_err(|e| format!("failed to persist secrets for {provider_id}: {e}"))?;

        saved_count += saved.len();
    }

    Ok(saved_count)
}
